bpo-30357 each test in test_thread waits until all spawned threads finish

grzgrzgrz3 opened the pull request.

grzgrzgrz3 committed:
bpo-30357 each test in test_thread waits until all spawn threads finish

vstinner commented on Lib/test/test_thread.py:
```diff
@@ -32,6 +32,11 @@ def setUp(self):
         self.next_ident = 0
+        self.thread_count = thread._count()
+
+    def tearDown(self):
```
This code looks similar to threading_cleanup(): please use threading_setup() and threading_cleanup() from test.support.

grzgrzgrz3 committed:
bpo-30357 each test in test_thread waits until all spawn threads finish

vstinner commented:
LGTM, but I would prefer to fix experiment this change in the master branch, before touching the super-stable 2.7 branch. Can you propose the same change for the master branch please?

grzgrzgrz3 commented:
I was unable to reproduce this issue on the master branch. However still it's worth to clean up tests correctly. I will create new PR for master branch.

What you think about rewriting `test_save_exception_state_on_error` and use `_thread._set_sentinel` instead `thread._count`?

vstinner commented:
You also need to add yourself to Misc/ACKS. You might also document the change in Misc/NEWS. I created https://github.com/python/cpython/pull/1592 for master where I credited you ;-)

vstinner commented:
> I was unable to reproduce this issue on the master branch. However still it's worth to clean up tests correctly.

Well, we are talking about race conditions. It's really hard to reproduce them. It's likely that the test order changed between Python 2.7 and master, but it doesn't mean that master is bugfix. Just that the bug is less likely :-)

> I will create new PR for master branch.

Sorry, I wanted to fix quickly the bug, so I pushed a fix based on your work on the master branch and added your name ;-) Can you please try to backport the fix to 2.7? Try to cherry-pick the change from master (or complete this change).

grzgrzgrz3 committed:
bpo-30357: test_thread now uses threading_cleanup() (#1592)

test_thread: setUp() now uses support.threading_setup() and
support.threading_cleanup() to wait until threads complete to avoid
random side effects on following tests.

Co-Authored-By: Victor Stinner <victor.stinner@gmail.com>

grzgrzgrz3 committed:
bpo-30357: test_thread now uses threading_cleanup() (#1592)

test_thread: setUp() now uses support.threading_setup() and
support.threading_cleanup() to wait until threads complete to avoid
random side effects on following tests.

Co-Authored-By: Victor Stinner <victor.stinner@gmail.com>

vstinner commented:
LGTM! Thanks for the backport.
